web-new/src/pages/ConfigEditor.tsx

```diff
@@ -1,10 +1,151 @@
+import useSWR from "swr";
+import * as monaco from "monaco-editor";
+import { configureMonacoYaml } from "monaco-yaml";
+import { useCallback, useEffect, useRef, useState } from "react";
+import { useApiHost } from "@/api";
 import Heading from "@/components/ui/heading";
+import ActivityIndicator from "@/components/ui/activity-indicator";
+import { Button } from "@/components/ui/button";
+import axios from "axios";
+import copy from "copy-to-clipboard";
+
+type SaveOptions = "saveonly" | "restart";
 
 function ConfigEditor() {
+  const apiHost = useApiHost();
+  const { data: config } = useSWR<string>("config/raw");
+  const [success, setSuccess] = useState<string | undefined>();
+  const [error, setError] = useState<string | undefined>();
+  const editorRef = useRef<any | null>();
+
+  const onHandleSaveConfig = useCallback(
+    async (save_option: SaveOptions) => {
+      if (!editorRef.current) {
+        return;
+      }
+
+      axios
+        .post(
+          `config/save?save_option=${save_option}`,
+          editorRef.current.getValue(),
+          {
+            headers: { "Content-Type": "text/plain" },
+          }
+        )
+        .then((response) => {
+          if (response.status === 200) {
+            setError("");
+            setSuccess(response.data.message);
+          }
+        })
+        .catch((error) => {
+          setSuccess("");
+
+          if (error.response) {
+            setError(error.response.data.message);
+          } else {
+            setError(error.message);
+          }
+        });
+    },
+    [editorRef]
+  );
+
+  const handleCopyConfig = useCallback(async () => {
+    if (!editorRef.current) {
+      return;
+    }
+
+    copy(editorRef.current.getValue());
+  }, [editorRef]);
+
+  useEffect(() => {
+    if (!config) {
+      return;
+    }
+
+    if ((document?.getElementById("container")?.children || []).length > 0) {
```

Comment:
Maybe this could be more react-like by having some kind of `editorInitialized` state and then checking that here.